sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java

```diff
@@ -191,10 +197,10 @@ private InProcessEvaluationContext(
       // Empty bundles don't impact watermarks and shouldn't trigger downstream execution, so
       // filter them out
       if (!Iterables.isEmpty(committed.getElements())) {
-        completed.add(committed);
+        outputs.put(committed, valueToConsumers.get(committed.getPCollection()));
```

Comment:
Since `committed` carries with it all the information needed for a lookup in `valueToConsumers`, perhaps you don't need to convert the iterable to a map so early? Can it be deferred just to the moment you add them as pending?